src/js/modules/supervisors/Repository.ts

```diff
@@ -87,31 +90,68 @@ class Repository {
     return this.handles.size;
   }
 
+  createEmptyProcessBatchReplay(
+    handle: Handle,
+    requestItem: ProcessBatchRequestItem
+  ): ProcessBatchReplyItem {
+    return {
+      coprocessorId: BigInt(handle.coprocessor.globalId),
+      ntp: {
+        ...requestItem.ntp,
+        topic: `${requestItem.ntp.topic}`,
+      },
+      resultRecordBatch: [],
+    };
+  }
+
   transformMapResultToArray(
     requestItem: ProcessBatchRequestItem,
     handle: Handle,
     recordBatch: RecordBatch,
     resultRecordBatch: Map<string, RecordBatch>
   ): ProcessBatchReplyItem[] {
     const results: ProcessBatchReplyItem[] = [];
-    for (const [key, value] of resultRecordBatch) {
-      value.records = value.records.map((record) => {
-        record.length = calculateRecordLength(record);
-        record.valueLen = record.value.length;
-        return record;
-      });
-      value.header.sizeBytes = calculateRecordBatchSize(value.records);
-      value.header.term = recordBatch.header.term;
-      results.push({
-        coprocessorId: BigInt(handle.coprocessor.globalId),
-        ntp: {
-          ...requestItem.ntp,
-          topic: `${requestItem.ntp.topic}.$${key}$`,
-        },
-        resultRecordBatch: [value],
-      });
-      return results;
+    if (resultRecordBatch === undefined) {
+      /*
+       Coprocessor doesn't return anything, in this case, it responses to
+       empty process batch replay to Redpanda in order to avoid, send this
+       request again.
+      */
+
+      this.logger.warn(
+        `coprocessor (${handle.coprocessor.globalId}) doesn't` +
+          `return a Map<string, RecordBatch>`
+      );
+      results.push(this.createEmptyProcessBatchReplay(handle, requestItem));
+    } else if (resultRecordBatch.size === 0) {
+      /*
+       Coprocessor returns a empty Map, in this case, it responses to
```

Comment:
Same suggestion above applies here

Reply:
I think this is different because this is an explicit empty map, that value is necessary for response to redpanda. also, the return type is Promise<Map>.